fix #1239 - Command-line based syntax errors now print location

baptistemanson opened the pull request:
Hi all!

I took the path to implement a custom errorHandler, to be more aligned with the rest of the code, but it may be a mistake.

I reached that point.
```bash
echo "(){}" | yarn prepack
yarn prepack v0.27.5
$ node lib/prepack-cli.js
Unexpected token (1:1) found in stdin
Error: A fatal error occurred while prepacking.
+ (full stack trace)
error Command failed with exit code 1.
```

and when srcmapIn is given but not existing:
```bash
 echo "(){}" | yarn prepack -- --srcmapIn notExisting
yarn prepack v0.27.5
$ node lib/prepack-cli.js "--srcmapIn" "notExisting"
No sourcemap found at notExisting.
Unexpected token (1:1) found in stdin
+ (stack trace)
```
Finally when using files with syntax error:
```bash
yarn prepack ../notepad/test.js
yarn prepack v0.27.5
$ node lib/prepack-cli.js "../notepad/test.js"
Unexpected token (1:3) found in input file ../notepad/test.js
+ (stack trace)
```

hermanventer commented:
We never want a stack trace unless there is an actual bug in Prepack.

hermanventer commented on src/prepack-node.js:
```diff
@@ -34,14 +34,21 @@ export function prepackStdin(options: PrepackOptions = defaultOptions, callback:
+              fileContents: code,
+              sourceMapContents: sourceMap,
+            },
+          ],
```
This reformatting seems unnecessary.

baptistemanson replied:
Prettier added it. I will manually correct it.

baptistemanson commented:
I will correct to remove the stack trace + failing test. Thank you for being patient with me!

baptistemanson commented:
- I manually reverted the prettier reformatting.
- I ran prettier though on the test files
- I inlined the syntax error file
- I fixed the unit tests

I sadly couldn't figure out how to make the difference between:
- a FatalError due to a user input issue, like a syntax error and 
- a FatalError coming from prepack itself which would require the complete stack trace.
So the stack trace is still present, with stdin or with a filename.

hermanventer commented:
A FatalError is the Prepack way of saying "an error has already been reported, now just stop running".

baptistemanson commented:
Thank you @hermanventer you are very patient with me.
- I undertand that I can catch at the highest level possible any "FatalError" .
- I need to make sure that we always called handleError with a CompilerDiagnostic before each potential FatalError throw.

I looked for a DEBUG flag/env var to bypass the catch, but couldn't find any yet.
So I thought I could check the status of errors[] array, and only catch if errors has at least one CompilerDiagnostic in it. What do you think?

hermanventer commented:
It would be a programming error if a FatalError reaches the final back stop without there being a diagnostic, so it would be appropriate to assert this with an invariant.

baptistemanson commented:
I think it is ready for review now. Have a nice weekend!

hermanventer commented on scripts/test-std-in.sh:
```diff
@@ -1,22 +1,26 @@
-cat ./test/std-in/StdInError.js | node ./bin/prepack.js --out StdInError-test.js 2> /dev/null
-if [[ $? -ne 1 ]]; then
+(echo "{}()" | node ./bin/prepack.js --out StdInError-test.js  2>&1 1>/dev/null ) | grep "In stdin:(1:4) FatalError PP1004: Syntax error: Unexpected token (1:3)" > /dev/null
+if [[ $? -ne 0 ]]; then
```
The expected result for this test is exit code 1, as it says in the comment.

baptistemanson replied:
Sure, I will separate it in two assertions, one for the message and one for the error.

hermanventer commented on src/prepack-cli.js:
```diff
@@ -227,16 +227,35 @@ function run(
+      invariant(errors.size > 0, "FatalError must generate at least one CompilerDiagnostic");
+    } else {
+      console.error(x.stack);
+      process.exit(1);
```
It would be better if any diagnostics that were collected were printed out before printing the stack trace and terminating the process. One way to do this it make the catch wrap the finally.

baptistemanson replied:
Yes.

hermanventer commented on src/prepack-cli.js:
```diff
@@ -227,16 +227,35 @@ function run(
         foundFatal = foundFatal || error.severity === "FatalError";
         console.error(
-          `${loc.source || ""}(${loc.start.line}:${loc.start.column +
+          `${sourceMessage}(${loc.start.line}:${loc.start.column +
```
In line 263 below throw a FatalError so that the catch clause (which should now be wrapping the finally), can terminate the process with exit code 1.

baptistemanson commented:
Hi @hermanventer,
- I split the unit tests in 2 checks: return signal and error message.
- I decided to create a printDiagnostics function.
- I deleted on purpose stderror, it was acting as a duplicate of echo "{}()" that you guys provided, and it triggers an error in most IDEs.

- I didn't double wrap try{}catch{}finally{} as you suggested, not that it was a bad idea ;). I just found the code readable enough when I introduced "printDiagnostics".
